internal/target/service_list_more.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package target
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/refreshtoken"
+	rt "github.com/hashicorp/boundary/internal/refreshtoken"
+)
+
+func ListMore(
+	ctx context.Context,
+	tok *rt.RefreshToken,
+	repo *Repository,
+	grantsHasher rt.GrantsHasher,
+	pageSize int,
+	filterItemFn func(Target) (bool, error),
+) (*ListResponse, error) {
+	const op = "target.ListMore"
+
+	deletedIds, transactionTimestamp, err := repo.ListDeletedIds(ctx, tok.CreatedTime)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	limit := pageSize + 1
+	opts := []Option{
+		WithLimit(limit),
+		WithStartPageAfterItem(tok.LastItemId, tok.LastItemUpdatedTime),
+	}
+
+	// pagination magic
+	targets := make([]Target, 0, pageSize+1)
+dbLoop:
+	for {
+		// Request another page from the DB until we fill the final items
+		page, err := repo.ListTargets(ctx, opts...)
+		if err != nil {
+			return nil, errors.Wrap(ctx, err, op)
+		}
+		for _, item := range page {
+			ok, err := filterItemFn(item)
+			if err != nil {
+				return nil, errors.Wrap(ctx, err, op)
+			}
+			if ok {
+				targets = append(targets, item)
+				// If we filled the items after filtering,
+				// we're done.
+				if len(targets) == cap(targets) {
+					break dbLoop
+				}
+			}
+		}
+		// If the current page was shorter than the limit, stop iterating
+		if len(page) < limit {
+			break dbLoop
+		}
+
+		opts = []Option{
+			WithLimit(limit),
+			WithStartPageAfterItem(page[len(page)-1].GetPublicId(), page[len(page)-1].GetUpdateTime().AsTime()),
+		}
+	}
+	// If we couldn't fill the items, it was a complete listing.
+	completeListing := len(targets) < cap(targets)
+	if !completeListing {
+		// Items is of size pageSize+1, so
+		// truncate if it was filled.
+		targets = targets[:pageSize]
+	}
+
+	totalItems, err := repo.EstimatedCount(ctx)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	grantsHash, err := grantsHasher.GrantsHash(ctx)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	resp := &ListResponse{
+		Items:               targets,
+		DeletedIds:          deletedIds,
+		EstimatedTotalItems: totalItems,
+		CompleteListing:     completeListing,
+		RefreshToken: &refreshtoken.RefreshToken{
+			// Use the timestamp of the deleted IDs transaction with a
+			// buffer to account for overlapping transactions. It is okay
+			// to return a deleted ID more than once. The buffer corresponds
+			// to Postgres' default transaction timeout.
+			CreatedTime:         transactionTimestamp.Add(-30 * time.Second),
```

Comment:
This advances the CreateTime of the refresh token, which means that, as long as refresh requests keep happening, the token will never expire. Is this the intended behavior? I thought refresh tokens are meant to eventual expire to force clients to make a fresh request.

For example, the only way [this condition would be true](https://github.com/hashicorp/boundary/pull/3817/files#diff-1811705c88d7ca2ba4ba4ac4bde5dce13d821453ce236c4bb834d3e11bfc1e85R59) is if a client takes more than 30 days to make a follow-up request.

Reply:
Yes, this is my understanding of the RFC, that tokens should only expire if not used for 30 days.

Reply:
My intention was to force a full refresh every 30 days.

Reply:
Seems I misunderstood this part of the RFC, thanks for the clarification. We'll need to track two timestamps then, the create time and "update" time. We'll make the changes necessary!

Reply:
We've added a `New`, `Refresh` and `RefreshLastItem` set of functions, please take a look.